parser/src/aspects/substitution.rs

```diff
@@ -16,14 +17,20 @@ impl<'a> SubstitutionAspect<'a> for Parser<'a> {
     fn substitution(&mut self) -> ParseResult<Expr<'a>> {
         let start_token = self.cursor.force(
             of_types(&[TokenType::At, TokenType::Dollar]),
-            "Expected at or dollar sign.",
+            "Expected '@' or '$' sign.",
         )?;
 
         // Short pass for variable references
-        if self
-            .cursor
-            .lookahead(of_type(TokenType::RoundedLeftBracket))
-            .is_none()
+        if self.cursor
+            .lookahead(
+                not(of_type(TokenType::RoundedLeftBracket))
+                    .and_then(
+                        like(TokenType::is_valid_var_ref_name)
+                            .or(of_type(CurlyLeftBracket))
+                            .or(eox())
+                    )
```

Comment:
The look ahead should not look so far. Invalid variable names would not have a meaningful message.